System/Metrics/Prometheus/Internal/State.hs

```diff
@@ -543,18 +655,26 @@ sampleState State{..} = do
       (\r -> M.map (M.map ($ r)) groupSamplerMetrics)
         <$> groupSampleAction
 
+    sampleUncheckedDynamicGroupSampler
+      :: UncheckedDynamicGroupSampler -> IO Sample
+    sampleUncheckedDynamicGroupSampler UncheckedDynamicGroupSampler{..} = do
+      (\r -> M.map (fmap ($ r)) uncheckedDynamicGroupSamplerMetrics)
+        <$> uncheckedDynamicGroupSampleAction
+
 -- | Test for equality ignoring `MetricId`s and `GroupId`s.
 --
 -- This test assumes that each `GroupSampler` in `stateGroups` is
 -- unique, which follows from the `State` invariants.
 functionallyEqual :: SampledState -> SampledState -> Bool
-functionallyEqual state1 state2 = withoutIds state1 == withoutIds state2
+functionallyEqual state1 state2 =
+  checkedStateWithoutIds state1 == checkedStateWithoutIds state2
+     && sampledStateUncheckedDynamicGroups state1 == sampledStateUncheckedDynamicGroups state2
   where
-    withoutIds
+    checkedStateWithoutIds
       :: SampledState
       -> M.Map Name
           (Help, M.Map Labels (Either Value (Maybe SampleWithoutHelp)))
-    withoutIds state =
+    checkedStateWithoutIds state =
       flip (M.map . second . M.map) (sampledStateMetrics state) $
         \(e, _version) -> case e of
           Left value -> Left value
```

Comment:
Thanks, Alex, I think it is okay wherever it is :sweat_smile: given we have already documented the use of function.

I thought that we should have all the tests in one place as it might be confusing for someone looking at a function defined only for testing something defined there. But nevermind, this was totally an optional suggestion, we can skip it.